### PR TITLE
services/horizon: Add prometheus metrics to database queries

### PR DESCRIPTION
**Pull request by paulbellamy**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Add prometheus metrics for query count and duration to horizon database calls. Turns out the metrics we had before were actually missing the ingest and core database clients as well, so I've added the same metrics to those too.

The big code changes here are:
- Actually use `db.SessionInterface`, instead of just using `*db.Session` everywhere. Nice cleanup anyway.
- Have to pass the `StateMiddleware` and `HistoryMiddleware` into each route individually, as they have to happen after routing. Big OOF.

### Why

So we can verify and really understand which endpoints are causing db load and why.

Fixes https://github.com/stellar/go/issues/3593

### Known limitations

- [x] TODO: Needs a bit more testing, like firing it up and seeing it work.
- I'm not sure if the register/unregister calls are in the right place. wdyt?
- @jacekn Do these metrics and labels look right?

### CODE REVIEW

**Comment by ire-and-curses** [comment]
Just want to confirm: are we getting raw numbers from these metrics? It would be great to be able to derive e.g. max and min as well as average for these calls.

**Comment by jacekn** [comment]
> @jacekn Do these metrics and labels look right?

Yeah they look good to me.
We're splitting one of them by `route` that that one will generate 225 new metrics so in total it should be 270 metrics which if my math is right. That's not too bad, it's less than 25% increase and those should be really useful

**Comment by jacekn** [comment]
> Just want to confirm: are we getting raw numbers from these metrics? It would be great to be able to derive e.g. max and min as well as average for these calls.

Averages we'll get out of the box - in prometheus we'll be able to divide total of all queries by number of queries to get average.
Min/max we won't get out of the box but it's doable too. We'll just need a small "recording rule" in prometheus (which is a common thing and very simple to set up). With the recording rule we'll be able to calculate min/max in any timerange we choose.

File: services/horizon/internal/db2/history/accounts.go [hunk 1]
```diff
@@ -190,7 +191,7 @@ func (q *Q) UpsertAccounts(ctx context.Context, accounts []xdr.LedgerEntry) erro
 		num_sponsoring = excluded.num_sponsoring`
 
 	_, err := q.ExecRaw(
-		ctx,
+		context.WithValue(ctx, &db.QueryTypeContextKey, "upsert"),
```
Comment (2opremio) [inline services/horizon/internal/db2/history/accounts.go:194]: Can we get a symbol for the uses of `"upsert"`?

File: services/horizon/internal/httpx/middleware.go [hunk 2]
```diff
@@ -141,16 +141,16 @@ func sanitizeMetricRoute(routePattern string) string {
 	route = strings.ReplaceAll(route, "\\", "\\\\")
 	route = strings.ReplaceAll(route, "\"", "\\\"")
 	route = strings.ReplaceAll(route, "\n", "\\n")
+	if route == "" {
+		// Can be empty when request did not reached the final route (ex. blocked by
```
Comment (2opremio) [inline services/horizon/internal/httpx/middleware.go:145]:
```suggestion
		// Can be empty when request did not reach the final route (ex. blocked by
```

File: services/horizon/internal/init.go [hunk 3]
```diff
@@ -44,22 +44,24 @@ func mustInitHorizonDB(app *App) {
 	}
 
 	app.historyQ = &history.Q{mustNewDBSession(
+		"history",
```
Comment (2opremio) [inline services/horizon/internal/init.go:47]: can we get symbols for the subsystems?

**Review by 2opremio — APPROVED** [review]

File: services/horizon/internal/init.go [hunk 4]
```diff
@@ -18,15 +18,15 @@ import (
...
+func mustNewDBSession(subsystem db.Subsystem, databaseURL string, maxIdle, maxOpen int, registry *prometheus.Registry) db.SessionInterface {
 	session, err := db.Open("postgres", databaseURL)
 	if err != nil {
 		log.Fatalf("cannot open Horizon DB: %v", err)
 	}
 
 	session.DB.SetMaxIdleConns(maxIdle)
 	session.DB.SetMaxOpenConns(maxOpen)
-	return session
+	return db.RegisterMetrics(session, "horizon", subsystem, registry)
```
Comment (bartekn) [inline services/horizon/internal/init.go:29]: I think it changes metrics' names. Can we revert it? It will break Grafana dashboard.

**Pull request merged**